sqlmesh/core/metric/definition.py

```diff
@@ -0,0 +1,193 @@
+            **kwargs,
+        }
+    )
+
+    metric._path = path
+
+    return metric
+
+
+def expand_metrics(metas: UniqueKeyDict[str, MetricMeta]) -> UniqueKeyDict[str, Metric]:
+    """Resolves all metas into standalone metrics."""
+    metrics: UniqueKeyDict[str, Metric] = UniqueKeyDict("metrics")
+
+    for name, meta in metas.items():
+        if name not in metrics:
+            metrics[name] = meta.to_metric(metas, metrics)
+
+    return metrics
+
+
+def find_tables(expression: exp.Expression, dialect: str) -> t.Set[str]:
+    """Finds all the table references in a metric definition."""
+    return {
+        d.normalize_model_name(exp.table_(*reversed(column.parts[:-1])), dialect=dialect)  # type: ignore
+        for column in expression.find_all(exp.Column)
+        if column.table
+    }
+
+
+@t.overload
+def remove_namespace(expression: str) -> str:
+    ...
+
+
+@t.overload
+def remove_namespace(expression: exp.Column, dialect: str) -> str:
+    ...
+
+
+def remove_namespace(expression: str | exp.Column, dialect: t.Optional[str] = None) -> str:
+    """Given a column or a string, rewrite table namespaces like catalog.db to catalog__db"""
+
+    if not isinstance(expression, str):
+        assert dialect is not None
+        expression = first(find_tables(expression, dialect=dialect))
+    return expression.replace(".", "__")
+
+
+class MetricMeta(PydanticModel, frozen=True):
+    """Raw metric definition without relationships or expansion of derived metrics."""
+
+    name: str
+    dialect: str
+    expression: exp.Expression
+    description: t.Optional[str] = None
+    owner: t.Optional[str] = None
+
+    _path: Path = Path()
+
+    @validator("name", pre=True)
+    def _name_validator(cls, v: t.Any, values: t.Dict[str, t.Any]) -> str:
+        return cls._string_validator(v).lower()
+
+    @validator("dialect", "owner", "description", pre=True)
+    def _string_validator(cls, v: t.Any) -> t.Optional[str]:
+        if isinstance(v, exp.Expression):
+            return v.name
+        return str(v) if v is not None else None
+
+    @validator("expression", pre=True)
+    def _validate_expression(
+        cls,
+        v: t.Any,
+        values: t.Dict[str, t.Any],
+    ) -> exp.Expression:
+        if isinstance(v, str):
+            return d.parse_one(v, dialect=values.get("dialect"))
+        if isinstance(v, exp.Expression):
+            return v
+        return v
+
+    def to_metric(
+        self, metas: t.Dict[str, MetricMeta], metrics: UniqueKeyDict[str, Metric]
+    ) -> Metric:
+        """Converts a metric meta into a fully expanded and standalone metric."""
+        payload = self.dict(exclude={"expression"})
+        metric_refs = {}
+
+        agg_or_ref = False
+
+        for node, *_ in self.expression.walk():
+            if isinstance(node, exp.Alias):
+                _raise_metric_config_error(
+                    f"Alias found for metric '{self.name}' which is not allowed", self._path
```

Comment:
When and why this might occur?

Reply:
user can write bad sql in the definition

Reply:
I don't think we protect against all sorts of ways a user can write bad sql in there. I'm just curious why this case was picked specifically.

Reply:
because the formulas have a / b as c and user may think this is how they should define their stuff and add an alias

Reply:
ah, I see